src/toil_scripts/gatk_germline/germline.py

```diff
@@ -1,688 +1,570 @@
+    if config.run_bwa:
+        bwa_references = {'amb', 'ann', 'bwt', 'pac', 'sa', 'alt'}
+        unrequired_references.add('alt')
+        references |= bwa_references
+    if config.preprocess:
+        preprocess_references = {'mills', 'phase', 'dbsnp'}
+        references |= preprocess_references
+    if config.run_vqsr:
+        vqsr_references = {'phase', 'mills', 'dbsnp', 'hapmap', 'omni'}
+        references |= vqsr_references
+    for name in references:
+        try:
+            url = getattr(config, name)
+            if url is None:
+                continue
+            setattr(config, name, job.addChildJobFn(download_url_job,
+                                                    url,
+                                                    name=name,
+                                                    s3_key_path=config.ssec).rv())
+        except AttributeError:
+            setattr(config, name, None)
+        finally:
+            if getattr(config, name) is None and name not in unrequired_references:
+                raise ValueError("Necessary parameter is missing:\n{}".format(name))
+    return job.addFollowOnJobFn(reference_preprocessing, config).rv()
+
+
+def reference_preprocessing(job, config):
+    """
+    Creates a genome.fa.fai and genome.dict file for reference genome.fa
+    if one is not already present in the config dictionary
+
+    :param JobFunctionWrapping Job: passed by Toil
+    :param Namespace config: Pipeline configuration options and shared files.
+                             Requires genome_fasta attribute
+    :return: Updated config with reference index files
+    :rtype: Namespace
+    """
+    job.fileStore.logToMaster('Preparing Reference Files')
+    genome_id = config.genome_fasta
+    if not hasattr(config, 'genome_fai') or config.genome_fai is None:
+        config.genome_fai = job.addChildJobFn(run_samtools_faidx, genome_id).rv()
+    if not hasattr(config, 'genome_dict') or config.genome_dict is None:
+        config.genome_dict = job.addChildJobFn(run_picard_create_sequence_dictionary,
+                                               genome_id).rv()
+    return config
+
+
+def prepare_bam(job, uuid, url, config, rg_line=None):
+    """
+    Prepares BAM file for GATK Germline pipeline.
+
+    0: Align FASTQ or Download BAM
+    1: Remove secondary alignments
+    2: Sort BAM
+    3: Index BAM
+
+    :param JobFunctionWrappingJob job: Toil Job instance
+    :param str uuid: Unique identifier for the sample
+    :param str url: URL to BAM file
+    :param Namespace config: Configuration options for pipeline
+    :param str rg_line: RG line for BWA alignment. Default is None.
+    :return: BAM and BAI FileStoreIDs
+    :rtype: tuple
+    """
+
+    cores = multiprocessing.cpu_count()
+
+    # Determine extension on input file
+    base, ext1 = os.path.splitext(url)
+    _, ext2 = os.path.splitext(base)
+
+    # Produce or download BAM file
+    if {ext1, ext2} & {'.fq', '.fastq'}:
```

Comment:
This syntax always looks odd to me. As an aside though, this should include the GZIPed variants of FASTQ.